<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  
  <meta name="generator" content="Hugo 0.72.0" />

  
  <meta name="description" content="Some description">
  

  
  <link rel="apple-touch-icon" sizes="180x180" href="https://memotut.com/apple-touch-icon.png">

  
  <link rel="icon" type="image/png" sizes="32x32" href="https://memotut.com/favicon-32x32.png">

  
  <link rel="icon" type="image/png" sizes="16x16" href="https://memotut.com/favicon-16x16.png">

  
  <link rel="manifest" href="https://memotut.com/site.webmanifest">

  
  <link rel="mask-icon" href="https://memotut.com/safari-pinned-tab.svg" color="#5bbad5">

  <meta name="msapplication-TileColor" content="#da532c">

  <meta name="theme-color" content="#ffffff">

  
  <link rel="stylesheet" href="https://memotut.com/css/bootstrap.min.css" />

  
  <title>Probabilistic model mmvec to detect correlation between microbiome and metabolome | Memo Tut</title>
  

  <style>
body {
  min-width: 300px;
}

.custom-navbar {
  margin-bottom: 1em;
  height: 60px;
}

.custom-navbar a {
  display: inline-block; 
  padding: 18px 0;
  margin-right: 1em; 
  font-weight: bold; 
}

.custom-navbar a:hover,
.custom-navbar a:focus {
  text-decoration: none; 
}

@media print {
  .custom-navbar {
    display: none;
  }
}

article {
  padding-bottom: 1em;
}

img {
  max-width: 100%;
}


body {
  background-color: #fff;
}



body {
  color: #212529;
}



a {
  color: #007bff;
}



a:hover,
a:focus {
  color: #0056b3;
}



.custom-navbar {
  background-color: #212529;
}



.custom-navbar a {
  color: rgba(255, 255, 255, 0.75);
}



.custom-navbar a:hover,
.custom-navbar a:focus {
  color: rgba(255, 255, 255, 1);
}



.container {
  max-width: 800px;
}



pre {
  display: block;
  padding: 9.5px;
  word-break: break-all;
  word-wrap: break-word;
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 4px;
}

pre code {
  padding: 0;
  font-size: inherit;
  color: inherit; 
  white-space: pre-wrap;
  background-color: transparent;
  border: none;
  border-radius: 0;
}

code {
  padding: 2px 4px;
  color: inherit; 
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 4px;
  font-size: .9em;
}



blockquote,
.blockquote {
  padding: 10px 20px;
  margin: 0 0 20px;
  font-size: 1em;
  border-left: 5px solid #6c757d;
}

</style>
</head>

<body>
  <nav class="custom-navbar">
  <div class="container">
    
    <a href="/">Posts</a>
    
    <a href="/tags/">Tags</a>
    
    <a href="/about/">About</a>
    
    <a href="/index.xml">RSS</a>
    
  </div>
</nav>
  
  <div class="container">
    <article>
      <h1>Probabilistic model mmvec to detect correlation between microbiome and metabolome</h1>
<p>
  <small class="text-secondary">
  
  
  Dec 25, 2019
  </small>
  

<small><code><a href="https://memotut.com/tags/python">Python</a></code></small>


<small><code><a href="https://memotut.com/tags/machine-learning"> machine learning</a></code></small>


<small><code><a href="https://memotut.com/tags/bioinformatics"> bioinformatics</a></code></small>


<small><code><a href="https://memotut.com/tags/metagenome"> metagenome</a></code></small>

</p>
<pre><code>Article on the last day of the Metagenomic Advent Calendar. I'm sorry for just registering and posting...
</code></pre>
<p>Introduction of how to use microbe–metabolite vectors, abbreviated <strong>mmvec</strong>, a method to detect the correlation between microbiome and metabolome.
<a href="https://www.nature.com/articles/s41592-019-0616-3">(Morton, et al. &ldquo;Learning representations of microbe–metabolite interactions.&rdquo; Nature Methods 16, 1306–1314 (2019))</a>
As you can see from the fact that word2vec and word sense are aligned (?), it is a method of learning latent vectors expressing the relationship between paired microbiome and metabolome data sets.</p>
<h2 id="problem-setting">Problem setting</h2>
<p>Correlation is very difficult in microbiome analysis. Correlation does not mean a cause and effect, but as a problem before that, the correlation that is generated by correlating microbiome data with other metadata or omics data is often not reliable.
The root problem is (often treated as) the microbiome is essentially <strong>compositional data</strong>(<a href="https://en.wikipedia.org/wiki/Compositional_data">Compositional data</a>).
It is known that when a simple correlation coefficient is calculated using data described as relative abundance such as lineage composition and gene functional composition, false positives are very likely to be detected.
For example, as a simple example, suppose that a time series analysis of microbiome changes associated with an increase in certain environmental metadata is performed. Suppose one line grows rapidly in response to increasing environmental metadata, while most other species grow slowly. Converting this data to relative abundances and calculating the Pearson and Spearman correlation coefficients, some strains grow positively in correlation with environmental metadata and most other strains negatively correlate with environmental metadata. It seems to decrease (because the sum in the composition data is competing for a certain constraint = 100% pie). As a result, a large amount of false positives will be detected.
This issue has been discussed repeatedly and repeatedly since the days of Pearson, and several countermeasures have been proposed. <a href="https://dx.plos.org/10.1371/journal.pcbi.1002687">SparCC</a>,<a href="https://dx.plos.org/10.1371/journal.pcbi.1004075">Proportionality</a>,<a href="https://dx.plos.org/10.1371/journal.pcbi.1004226">SPIEC-EASI</a> etc.
However, methods such as SparCC that depend on the log ratio transformation of abundance are not problematic to calculate in a single data set, such as when performing correlation analysis only in microbiome data, but the total count If the calculation is done between *<em>of</em> multiple data sets with significantly different values, the assumption of log ratio conversion does not hold and the calculation is not valid. The difference in scale between Meta16S and Fungal ITS is still good, but between Meta16S and Metabolome, the scale is so different that it becomes a serious problem.
The interaction between microbiome and metabolome is very important as a research subject, but the &ldquo;correlation&rdquo; cannot be calculated correctly. Then, this paper deals with what to do.</p>
<p>*Note that there are still discussions about whether this method has a higher detection accuracy than existing methods.
<a href="https://www.biorxiv.org/content/10.1101/847475v1">Rebuttal</a>(ClaimthatlinearmethodsuchasPearsonismuchmoreaccuratethanmmvecifconvertedappropriately)
<a href="https://www.biorxiv.org/content/10.1101/2019.12.10.871905v1">Rebuttal to the objection</a>(ThedatasetI&rsquo;mtestingisuseless,theactualdataisverysparse,andaftertestingwithsuchdatammvecisWin,andobjection)</p>
<h2 id="mmvec-algorithm">mmvec algorithm</h2>
<p>The concept of not calculating the &ldquo;correlation&rdquo; but estimating the conditional probability $P(\nu|\mu)$ of the metabolite ($\nu$) when the microorganism ($\mu$) is present ..</p>
<img width='600' src='https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/176102/bb3e18e0-3a39-55bf-4124-010b5be06c78.png' align= 'left'>
<p>The above figure is from <a href="https://github.com/biocore/mmvec">mmvec repository</a>.
This figure is a little misleading and looks like an auto encoder or VAE, but in reality it is just a simple stochastic model MAP estimation with tensorflow. Somehow, I&rsquo;m arguing about things like neural networks, but I&rsquo;m not sure if this is a neural network model (I can not say because I do not know the definition of NN, but if this is NN I think that logistic regression is also NN)</p>
<h3 id="generative-model">generative model</h3>
<p>Normal distribution for each of microorganisms \mu \in \{1, &hellip;N\}$ and metabolites $\nu \in \{1, &hellip;M\}$ as follows Generate p-dimensional microbial vector $u_\mu$ and metabolite vector $v_\nu$ from. (Low-dimensional p is set by the user)
$$u_\mu \sim \mathcal{N}(0, \sigma_\mu I)$$
$$v_\nu \sim \mathcal{N}(0, \sigma_\nu I)$$
For the phylogenetic annotation $\mu$ of the sequence sampled from the microbiome data, the conditional probability of the metabolite $v$,
$$P(v|\mu) = \frac {exp \left( v_\nu \cdot u_\mu + v_{\nu0} + u_{\mu0} \right)} {\sum_j exp \left( v_j \ cdot u_\mu + v_{j0} + u_{\mu0} \right)}$$
And The abundance of metabolites $y_k$ (total $n$) in sample $k$ is sampled from this multinomial distribution $q$,
$$q = [P(v_1|\mu), &hellip;P(v_M|\mu)]$$
$$y_k \sim Multinomial(n, q)$$
And The above simple model.
This is a table (<a href="https://en.wikipedia.com">Center log ratio conversion</a>ofconditionalprobability$P(\nu|\mu)$ofmetabolite$\nu$foreachmicroorganism$\mu$.org/wiki/Compositional_data#Center_logratio_transform)) is decomposed into a matrix $U$ in which microorganism vectors are arranged and a matrix $V$ in which metabolite vectors are arranged.
The inference uses the adam optimizer of tensorflow to maximize the following log-likelihood. If the array set of sample $k$ is $x_k$,
$$L = L_y + L_U + L_V$$
$$L_y = \sum_k \sum_{r \in x_k} ln Multinomial(y_k | q)$$
$$L_U = \sum_\mu \sum_{p=1}^{p} ln \mathcal{N}(U_{\mu,p} | 0, \sigma_\mu)$$
$$L_V = \sum_\nu \sum_{p=1}^{p} ln \mathcal{N}(V_{\nu,p} | 0, \sigma_\nu)$$</p>
<p>Looking at the calculation graph with TensorBoard, it looks like ↓.</p>
<img width='900' src='https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/176102/a3394629-de5f-ee2f-afee-a877e2cf2f0b.png' align= 'left'>
<p>The log-likelihood is calculated for all the individual sequences in the microbiome data, and the parameters are updated.
According to the description in the paper method, both the $u$ vector and the $v$ vector learned in this way can be directly converted into p-dimensional ratio vectors (which can be regarded as the center logratio conversion from Aitchison simplex). It seems that it can be interpreted in the same way as the composition of topics in.</p>
<h2 id="apply-to-soil-microbial-data">Apply to soil microbial data</h2>
<p>Installation is simply <code>pip install mmvec</code> etc.
It takes a long time to learn, so put <code>tensorflow-gpu </code> if you use GPU to calculate.
There seems to be a way to integrate it into QIIME2 as a plugin, but I&rsquo;m not sure because I haven&rsquo;t used QIIME. It seems that it is better to use the standalone version because it seems that GPU can not be used or there are restrictions.</p>
<p>The subject matter is metagenomic and metabolomic data of soil microbial communities, which is also dealt with in Fig. 3 of <a href="https://www.nature.com/articles/s41592-019-0616-3">mmvec paper</a>.
<a href="https://www.nature.com/articles/s41467-017-02356-9">Swenson, et al. &ldquo;Linking soil biology and chemistry in biological soil crust using isolate exometabolomics.&rdquo; Nature Communications 9, 19 (2018)</a>
Try to reproduce Fig.3a of mmvec paper.</p>
<p>First, read the data.
Thankfully, Soil Microbial Papers are Supplementary Data 1 and Supplementary Data 4 with 19 samples of <a href="https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-017-02356-9/MediaObjects/41467_2017_2356_MOESM7_ESM.xlsx">strain composition</a>,<a href="https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-017-02356-9/MediaObjects/41467_2017_2356_MOESM4_ESM.xlsx">Metabolitecomposition</a> Since it is distributed as an Excel file, I load it as it is, format it a little and save it as a tsv table file.
First, microbiome data.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#f92672">import</span> pandas <span style="color:#f92672">as</span> pd
microbe <span style="color:#f92672">=</span> pd<span style="color:#f92672">.</span>read_excel(<span style="color:#e6db74">&#39;./data/41467_2017_2356_MOESM7_ESM.xlsx&#39;</span>, \skiprows<span style="color:#f92672">=</span><span style="color:#ae81ff">0</span>, header<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>, index_col<span style="color:#f92672">=</span><span style="color:#ae81ff">0</span>)
microbe <span style="color:#f92672">=</span> microbe<span style="color:#f92672">.</span>drop([<span style="color:#e6db74">&#39;phylum&#39;</span>, <span style="color:#e6db74">&#39;class&#39;</span>, <span style="color:#e6db74">&#39;order&#39;</span>, <span style="color:#e6db74">&#39;family&#39;</span>, <span style="color:#e6db74">&#39;genus&#39;</span>, <span style="color:#e6db74">&#39;species&#39;</span>], axis<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>)
microbe <span style="color:#f92672">=</span> microbe<span style="color:#f92672">.</span>drop(<span style="color:#e6db74">&#39;9hr_late&#39;</span>, axis<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>) <span style="color:#75715e"># No corresponding sample in metabolites</span>
microbe <span style="color:#f92672">=</span> microbe <span style="color:#f92672">*</span> <span style="color:#ae81ff">1000.</span>
microbe<span style="color:#f92672">.</span>to_csv(<span style="color:#e6db74">&#39;./data/microbe.tsv&#39;</span>, sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;</span><span style="color:#ae81ff">\t</span><span style="color:#e6db74">&#39;</span>)
<span style="color:#66d9ef">print</span>(microbe<span style="color:#f92672">.</span>shape)
<span style="color:#66d9ef">print</span>(microbe<span style="color:#f92672">.</span>iloc[:<span style="color:#ae81ff">3</span>, :<span style="color:#ae81ff">3</span>])
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">    <span style="color:#f92672">(</span>468, 19<span style="color:#f92672">)</span>
                               3min_early  3min_earlymid  3min_latemid
    rplO ID                                                           
    rplo <span style="color:#ae81ff">1</span> <span style="color:#f92672">(</span>Cyanobacteria<span style="color:#f92672">)</span>         4121.0         4489.0        2032.0
    rplo <span style="color:#ae81ff">10</span> <span style="color:#f92672">(</span>Firmicutes<span style="color:#f92672">)</span>              0.0            0.0           0.0
    rplo <span style="color:#ae81ff">100</span> <span style="color:#f92672">(</span>Proteobacteria<span style="color:#f92672">)</span>       132.0           56.0          16.0
</code></pre></div><p>19サンプル、検出された系統は468。<br>
次にメタボロームデータ。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python">metabolite <span style="color:#f92672">=</span> pd<span style="color:#f92672">.</span>read_excel(<span style="color:#e6db74">&#39;./data/41467_2017_2356_MOESM4_ESM.xlsx&#39;</span>, \
                           skiprows<span style="color:#f92672">=</span><span style="color:#ae81ff">0</span>, header<span style="color:#f92672">=</span><span style="color:#ae81ff">1</span>, index_col<span style="color:#f92672">=</span><span style="color:#ae81ff">0</span>, \
                           usecols<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;A:AG&#39;</span>)
metabolite <span style="color:#f92672">=</span> metabolite<span style="color:#f92672">.</span>iloc[:, <span style="color:#ae81ff">13</span>:]
<span style="color:#75715e"># 微妙にサンプルのラベルが違うので、マイクロバイオームデータに合わせる</span>
metabolite <span style="color:#f92672">=</span> metabolite<span style="color:#f92672">.</span>rename(columns<span style="color:#f92672">=</span><span style="color:#66d9ef">lambda</span> l: l<span style="color:#f92672">.</span>split(<span style="color:#e6db74">&#39;_&#39;</span>)[<span style="color:#ae81ff">0</span>] <span style="color:#f92672">+</span><span style="color:#e6db74">&#39;_&#39;</span><span style="color:#f92672">+</span> l<span style="color:#f92672">.</span>split(<span style="color:#e6db74">&#39;_&#39;</span>)[<span style="color:#ae81ff">2</span>])
metabolite <span style="color:#f92672">=</span> metabolite<span style="color:#f92672">.</span>loc[:, microbe<span style="color:#f92672">.</span>columns]
metabolite<span style="color:#f92672">.</span>to_csv(<span style="color:#e6db74">&#39;./data/metabolite.tsv&#39;</span>, sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;</span><span style="color:#ae81ff">\t</span><span style="color:#e6db74">&#39;</span>)
<span style="color:#66d9ef">print</span>(metabolite<span style="color:#f92672">.</span>shape)
<span style="color:#66d9ef">print</span>(metabolite<span style="color:#f92672">.</span>iloc[:<span style="color:#ae81ff">3</span>, :<span style="color:#ae81ff">3</span>])
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">    <span style="color:#f92672">(</span>85, 19<span style="color:#f92672">)</span>
                                       3min_early  3min_earlymid  3min_latemid
    Name                                                                      
    <span style="color:#f92672">(</span>2,3-dihydroxy-3-methylbutanoate<span style="color:#f92672">)</span>        1.00   4.609149e+02  1.379355e+02
    <span style="color:#f92672">(</span>2,5-diaminohexanoate<span style="color:#f92672">)</span>             2782242.25   1.883228e+06  2.531814e+06
    <span style="color:#f92672">(</span>3-hydroxypyridine<span style="color:#f92672">)</span>                2764132.25   3.014544e+06  3.817224e+06
</code></pre></div><p>19サンプル、検出している代謝物質は85種類。</p>
<p>この２つのファイルを入力として、<code>mmvec paired-omics</code>コマンドで学習を実行する。
以下、オプション：
<code>--microbe-file, --metabolite-file</code>: 系統組成データ、メタボロームデータ。リポジトリのREADMEでは<a href="http://biom-format.org">biom-format</a>のファイルしか入力できないっぽく書かれてるけど、わざわざbiomに変換しなくても普通のタブ区切りテキストファイルを指定して実行できる。
<code>--summary-dir</code>: 結果を出力するディレクトリ
<code>--epochs</code>: エポック数
<code>--batch-size</code>: ミニバッチのサイズ
<code>--latent-dim</code>: 潜在ベクトルの次元
<code>--input-prior</code>: 事前分布のハイパーパラメータ$\sigma_\mu$
<code>--output-prior</code>: 事前分布のハイパーパラメータ$\sigma_\nu$
<code>--arm-the-gpu</code>: GPU使う場合はオンにする
<code>--learning-rate, --beta1, --beta2</code>: 学習率、Adamのパラメータ</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">mmvec paired-omics <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --microbe-file ./data/microbe.tsv <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --metabolite-file ./data/metabolite.tsv <span style="color:#ae81ff">\
</span><span style="color:#ae81ff"></span>    --summary-dir ./summary
</code></pre></div><p>エポックをどの程度にすればいいかは、データセットによって大きく異なるので一概には言えない。<br>
さいわい、RMSEや対数尤度を記録してくれるので、実行した場所で<code>tensorboard --logdir .</code>すればTensorBoard上で収束の確認はできる。今回の場合↓</p>
<img width='600' src='https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/176102/8159e820-a023-4f77-bea4-5d2f4a549d80.png' align='left'> 
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">ls summary
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">    latent_dim_3_input_prior_1.00_output_prior_1.00_beta1_0.90_beta2_0.95
    latent_dim_3_input_prior_1.00_output_prior_1.00_beta1_0.90_beta2_0.95_embedding.txt
    latent_dim_3_input_prior_1.00_output_prior_1.00_beta1_0.90_beta2_0.95_ordination.txt
    latent_dim_3_input_prior_1.00_output_prior_1.00_beta1_0.90_beta2_0.95_ranks.txt
</code></pre></div><p>学習が終わると3種類のファイルを出力する。
それぞれのファイルがなんなのか、ろくに説明がないのでわかりにくいが、ソースコードを読む限り以下のような結果が書き出されているらしい。
** XXX_embedding.txt
- それぞれの微生物、代謝物ごとの潜在ベクトル（u, v）</p>
<ul>
<li>XXX_ranks.txt
<ul>
<li>行列U, 行列Vを掛け算して中心化したマトリックス。条件付き確率$P(\nu|\mu)$をCenter logratio変換した値と一致。</li>
</ul>
</li>
<li>XXX_ordination.txt
<ul>
<li>上のranksファイルのマトリックスを特異値分解（u, s, v）して計算した、微生物の埋め込み（dot(u,s)）と代謝物の埋め込み（v）。論文中のbiplotはこの埋め込みで描かれている。</li>
</ul>
</li>
</ul>
<p>たぶん一番使い勝手がいいのはXXX_ranks.txtファイルで、普通の相関分析と同じように扱える。ただし、負の値は負の相関を意味するわけではなく、「全代謝物質の平均的な条件付き確率よりも下回っている」ことを意味している点に注意。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python">rankfile <span style="color:#f92672">=</span> <span style="color:#e6db74">&#39;./summary/latent_dim_3_input_prior_1.00_output_prior_1.00_beta1_0.90_beta2_0.95_ranks.txt&#39;</span>
ranks <span style="color:#f92672">=</span> pd<span style="color:#f92672">.</span>read_csv(rankfile, sep<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;</span><span style="color:#ae81ff">\t</span><span style="color:#e6db74">&#39;</span>, index_col<span style="color:#f92672">=</span><span style="color:#ae81ff">0</span>)
<span style="color:#66d9ef">print</span>(ranks<span style="color:#f92672">.</span>shape)
<span style="color:#66d9ef">print</span>(ranks<span style="color:#f92672">.</span>iloc[:<span style="color:#ae81ff">3</span>, :<span style="color:#ae81ff">3</span>])
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">    <span style="color:#f92672">(</span>85, 191<span style="color:#f92672">)</span>
                                       rplo <span style="color:#ae81ff">1</span> <span style="color:#f92672">(</span>Cyanobacteria<span style="color:#f92672">)</span>  rplo <span style="color:#ae81ff">100</span> <span style="color:#f92672">(</span>Proteobacteria<span style="color:#f92672">)</span>  rplo <span style="color:#ae81ff">101</span> <span style="color:#f92672">(</span>Proteobacteria<span style="color:#f92672">)</span>
    featureid                                                                                                      
    <span style="color:#f92672">(</span>2,3-dihydroxy-3-methylbutanoate<span style="color:#f92672">)</span>               -5.076497                  -0.229192                  -3.435717
    <span style="color:#f92672">(</span>2,5-diaminohexanoate<span style="color:#f92672">)</span>                          -1.053542                  -1.119804                  -0.508942
    <span style="color:#f92672">(</span>3-hydroxypyridine<span style="color:#f92672">)</span>                              0.055645                   0.001804                   0.797657
</code></pre></div><p>ここでは（mmvec論文同様）、<em>Microcoleus vaginatus</em> (&ldquo;rplo 1 (Cyanobacteria)&quot;) の結果に焦点をあてる。
もとの土壌微生物論文では、<em>Microcoleus vaginatus</em>の分離株の実験で、以下の代謝物を産生・放出することを確認していた。
しかし、メタゲノムのSpearman相関分析の結果、この中の一部については正の相関係数が検出されたが、半分は負の相関となってしまい、分離株と環境中で矛盾した結果となってしまった（<a href="https://www.nature.com/articles/s41467-017-02356-9">論文</a>のFig.3）。</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python">true_metabolites_text_offsets <span style="color:#f92672">=</span> {
    <span style="color:#e6db74">&#39;(3-methyladenine)&#39;</span>: (<span style="color:#f92672">-</span><span style="color:#ae81ff">0.</span>, <span style="color:#f92672">-</span><span style="color:#ae81ff">0.15</span>),
    <span style="color:#e6db74">&#39;7-methyladenine&#39;</span>: (<span style="color:#ae81ff">0.7</span>, <span style="color:#f92672">-</span><span style="color:#ae81ff">0</span>),
    <span style="color:#e6db74">&#39;4-guanidinobutanoate&#39;</span>: (<span style="color:#f92672">-</span><span style="color:#ae81ff">1.9</span>, <span style="color:#f92672">-</span><span style="color:#ae81ff">0.2</span>),
    <span style="color:#e6db74">&#39;uracil&#39;</span>: (<span style="color:#ae81ff">0.6</span>, <span style="color:#f92672">-</span><span style="color:#ae81ff">0.1</span>),
    <span style="color:#e6db74">&#39;xanthine&#39;</span>: (<span style="color:#ae81ff">0</span>, <span style="color:#ae81ff">0.1</span>),
    <span style="color:#e6db74">&#39;hypoxanthine&#39;</span>: (<span style="color:#ae81ff">0.2</span>, <span style="color:#f92672">-</span><span style="color:#ae81ff">0.2</span>),
    <span style="color:#e6db74">&#39;(N6-acetyl-lysine)&#39;</span>: (<span style="color:#f92672">-</span><span style="color:#ae81ff">1.7</span>, <span style="color:#ae81ff">0.1</span>),<span style="color:#e6db74">&#39;cytosine&#39;</span>: (<span style="color:#ae81ff">0</span>, <span style="color:#ae81ff">0.1</span>),
    <span style="color:#e6db74">&#39;N-acetylornithine&#39;</span>: (<span style="color:#ae81ff">0.2</span>, <span style="color:#ae81ff">0.1</span>),
    <span style="color:#e6db74">&#39;succinate&#39;</span>: (<span style="color:#f92672">-</span><span style="color:#ae81ff">0.5</span>, <span style="color:#ae81ff">0.2</span>),
    <span style="color:#e6db74">&#39;adenosine&#39;</span>: (<span style="color:#f92672">-</span><span style="color:#ae81ff">1.2</span>, <span style="color:#ae81ff">0.1</span>),
    <span style="color:#e6db74">&#39;guanine&#39;</span>: (<span style="color:#ae81ff">0.4</span>, <span style="color:#f92672">-</span><span style="color:#ae81ff">0.2</span>),
    <span style="color:#e6db74">&#39;adenine&#39;</span>: (<span style="color:#f92672">-</span><span style="color:#ae81ff">0.7</span>, <span style="color:#f92672">-</span><span style="color:#ae81ff">0.2</span>)}
</code></pre></div><p>For these metabolites, let&rsquo;s examine what the values are for the estimation of conditional probability by mmvec.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python">target <span style="color:#f92672">=</span><span style="color:#e6db74">&#39;rplo 1 (Cyanobacteria)&#39;</span> <span style="color:#75715e"># Microcoleus vaginatus</span>
Microcoleus <span style="color:#f92672">=</span> pd<span style="color:#f92672">.</span>DataFrame(ranks<span style="color:#f92672">.</span>loc[:, target]<span style="color:#f92672">.</span>values, \
                           index<span style="color:#f92672">=</span>ranks<span style="color:#f92672">.</span>index, columns<span style="color:#f92672">=</span>[<span style="color:#e6db74">&#39;log CondProb&#39;</span>])
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#f92672">%</span>matplotlib inline
<span style="color:#f92672">import</span> matplotlib.pyplot <span style="color:#f92672">as</span> plt
<span style="color:#f92672">import</span> seaborn <span style="color:#f92672">as</span> sns

df <span style="color:#f92672">=</span> Microcoleus<span style="color:#f92672">.</span>sort_values(by<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;log CondProb&#39;</span>, ascending<span style="color:#f92672">=</span>False)
colors <span style="color:#f92672">=</span> [<span style="color:#e6db74">&#39;red&#39;</span> <span style="color:#66d9ef">if</span> m <span style="color:#f92672">in</span> true_metabolites_text_offsets<span style="color:#f92672">.</span>keys() <span style="color:#66d9ef">else</span><span style="color:#e6db74">&#39;gray&#39;</span> <span style="color:#66d9ef">for</span> m <span style="color:#f92672">in</span> df<span style="color:#f92672">.</span>index]
fig, ax <span style="color:#f92672">=</span> plt<span style="color:#f92672">.</span>subplots(figsize<span style="color:#f92672">=</span>(<span style="color:#ae81ff">16</span>, <span style="color:#ae81ff">6</span>))
g <span style="color:#f92672">=</span> sns<span style="color:#f92672">.</span>barplot(x<span style="color:#f92672">=</span>df<span style="color:#f92672">.</span>index, y<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;log CondProb&#39;</span>, \
                palette<span style="color:#f92672">=</span>colors, data<span style="color:#f92672">=</span>df, ax<span style="color:#f92672">=</span>ax)
g<span style="color:#f92672">.</span>set_xticklabels(g<span style="color:#f92672">.</span>get_xticklabels(), rotation<span style="color:#f92672">=</span><span style="color:#ae81ff">90</span>)
ax<span style="color:#f92672">.</span>set_ylabel(<span style="color:#e6db74">&#39;log conditional probabilities&#39;</span>)
sns<span style="color:#f92672">.</span>despine()
</code></pre></div><img width="953" alt="output_29_0.png" src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/176102/7b32ccb5-f76a-107e-24b0-fc45438e3969.png">
<p>The red color is a metabolite that should be released by <em>Microcoleus vaginatus</em>.
It was found that the estimated values by mmvec were all positive values = probability values above the mean, including metabolites that had a negative Spearman correlation in the original paper.
Calculate Spearman correlations for each metabolite and compare the results.</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python"><span style="color:#f92672">import</span> scipy.stats

Microcoleus[<span style="color:#e6db74">&#39;Spearman rho&#39;</span>] <span style="color:#f92672">=</span> np<span style="color:#f92672">.</span>zeros(len(Microcoleus))[:, np<span style="color:#f92672">.</span>newaxis]
Microcoleus[<span style="color:#e6db74">&#39;P values&#39;</span>] <span style="color:#f92672">=</span> np<span style="color:#f92672">.</span>zeros(len(Microcoleus))[:, np<span style="color:#f92672">.</span>newaxis]

<span style="color:#66d9ef">for</span> m <span style="color:#f92672">in</span> Microcoleus<span style="color:#f92672">.</span>index:
    rho, p <span style="color:#f92672">=</span> scipy<span style="color:#f92672">.</span>stats<span style="color:#f92672">.</span>spearmanr(microbe<span style="color:#f92672">.</span>loc[target, :]<span style="color:#f92672">.</span>values, \
                                   metabolite<span style="color:#f92672">.</span>loc[m, :]<span style="color:#f92672">.</span>values)
    Microcoleus<span style="color:#f92672">.</span>loc[m,<span style="color:#e6db74">&#39;Spearman rho&#39;</span>] <span style="color:#f92672">=</span> rho
    Microcoleus<span style="color:#f92672">.</span>loc[m,<span style="color:#e6db74">&#39;P values&#39;</span>] <span style="color:#f92672">=</span> p
</code></pre></div><div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-python" data-lang="python">fig, ax <span style="color:#f92672">=</span> plt<span style="color:#f92672">.</span>subplots(figsize<span style="color:#f92672">=</span>(<span style="color:#ae81ff">12</span>, <span style="color:#ae81ff">12</span>))
ax<span style="color:#f92672">.</span>scatter(Microcoleus[<span style="color:#e6db74">&#39;log CondProb&#39;</span>], Microcoleus[<span style="color:#e6db74">&#39;Spearman rho&#39;</span>], \
          s<span style="color:#f92672">=-</span><span style="color:#ae81ff">10</span><span style="color:#f92672">*</span>np<span style="color:#f92672">.</span>log(Microcoleus[<span style="color:#e6db74">&#39;P values&#39;</span>]<span style="color:#f92672">.</span>values))
<span style="color:#66d9ef">for</span> m <span style="color:#f92672">in</span> true_metabolites_text_offsets<span style="color:#f92672">.</span>keys():
    x, y, _ <span style="color:#f92672">=</span> Microcoleus<span style="color:#f92672">.</span>loc[m, :]<span style="color:#f92672">.</span>values
    dx, dy <span style="color:#f92672">=</span> true_metabolites_text_offsets[m]
    fc <span style="color:#f92672">=</span><span style="color:#e6db74">&#39;red&#39;</span> <span style="color:#66d9ef">if</span> y <span style="color:#f92672">&lt;</span><span style="color:#ae81ff">0</span> <span style="color:#66d9ef">else</span><span style="color:#e6db74">&#39;black&#39;</span>
    ax<span style="color:#f92672">.</span>annotate(m, xy<span style="color:#f92672">=</span>(x, y), xytext<span style="color:#f92672">=</span>(x<span style="color:#f92672">+</span>dx, y<span style="color:#f92672">+</span>dy), fontsize<span style="color:#f92672">=</span><span style="color:#ae81ff">16</span>, \
                arrowprops<span style="color:#f92672">=</span>dict(facecolor<span style="color:#f92672">=</span>fc, shrink<span style="color:#f92672">=</span><span style="color:#ae81ff">0.01</span>))
ax<span style="color:#f92672">.</span>axhline(y<span style="color:#f92672">=</span><span style="color:#ae81ff">0</span>, color<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;gray&#39;</span>, linestyle<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;:&#39;</span>)
ax<span style="color:#f92672">.</span>axvline(x<span style="color:#f92672">=</span><span style="color:#ae81ff">0</span>, color<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;gray&#39;</span>, linestyle<span style="color:#f92672">=</span><span style="color:#e6db74">&#39;:&#39;</span>)
ax<span style="color:#f92672">.</span>set_xlabel(<span style="color:#e6db74">&#39;log conditional probabilities&#39;</span>, fontsize<span style="color:#f92672">=</span><span style="color:#ae81ff">18</span>)
ax<span style="color:#f92672">.</span>set_ylabel(<span style="color:#e6db74">&#39;Spearman correlations&#39;</span>, fontsize<span style="color:#f92672">=</span><span style="color:#ae81ff">18</span>)
plt<span style="color:#f92672">.</span>show()
</code></pre></div><img width="750" alt="output_32_0.png" src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/176102/3778396b-2d06-b858-0230-0dde21d6b5e2.png">
<p>The size of the circle indicates the p-value of Spearman correlation. The arrows indicate the metabolites that <em>Microcoleus vaginatus</em> should release (which should have a positive correlation). Red arrows indicate metabolites that are inconsistent with the isolate results in Spearman correlation.
As a result, it can be seen that the Spearman correlation has calculated negative correlation coefficients for some metabolites, but the calculation of mmvec can be calculated as a positive value.
It is difficult to define what the &ldquo;correct answer&rdquo; is and how to evaluate the effectiveness of the method is still under debate, but the results suggest that it looks like a promising approach.</p>

    </article>
  </div>

  
  
  
  <script>
  window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
  ga('create', 'UA-169005401-1', 'auto');
  ga('send', 'pageview');
  </script>
  <script async src='https://www.google-analytics.com/analytics.js'></script>
  

  

  
<link rel="stylesheet" type="text/css" href="//cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.1.0/cookieconsent.min.css" />
<script src="//cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.1.0/cookieconsent.min.js"></script>
<script>
window.addEventListener("load", function(){
window.cookieconsent.initialise({
  "palette": {
    "popup": {
      "background": "#216942",
      "text": "#b2d192"
    },
    "button": {
      "background": "#afed71"
    }
  }
})});
</script>

</body>

</html>
