<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  
  <meta name="generator" content="Hugo 0.72.0" />

  
  <meta name="description" content="Some description">
  

  
  <link rel="apple-touch-icon" sizes="180x180" href="https://memotut.com/apple-touch-icon.png">

  
  <link rel="icon" type="image/png" sizes="32x32" href="https://memotut.com/favicon-32x32.png">

  
  <link rel="icon" type="image/png" sizes="16x16" href="https://memotut.com/favicon-16x16.png">

  
  <link rel="manifest" href="https://memotut.com/site.webmanifest">

  
  <link rel="mask-icon" href="https://memotut.com/safari-pinned-tab.svg" color="#5bbad5">

  <meta name="msapplication-TileColor" content="#da532c">

  <meta name="theme-color" content="#ffffff">

  
  <link rel="stylesheet" href="https://memotut.com/css/bootstrap.min.css" />

  
  <title>[Python] Super easy molecular phylogenetic tree creation technique that I do not want to teach anyone | Memo Tut</title>
  

  <style>
body {
  min-width: 300px;
}

.custom-navbar {
  margin-bottom: 1em;
  height: 60px;
}

.custom-navbar a {
  display: inline-block; 
  padding: 18px 0;
  margin-right: 1em; 
  font-weight: bold; 
}

.custom-navbar a:hover,
.custom-navbar a:focus {
  text-decoration: none; 
}

@media print {
  .custom-navbar {
    display: none;
  }
}

article {
  padding-bottom: 1em;
}

img {
  max-width: 100%;
}


body {
  background-color: #fff;
}



body {
  color: #212529;
}



a {
  color: #007bff;
}



a:hover,
a:focus {
  color: #0056b3;
}



.custom-navbar {
  background-color: #212529;
}



.custom-navbar a {
  color: rgba(255, 255, 255, 0.75);
}



.custom-navbar a:hover,
.custom-navbar a:focus {
  color: rgba(255, 255, 255, 1);
}



.container {
  max-width: 800px;
}



pre {
  display: block;
  padding: 9.5px;
  word-break: break-all;
  word-wrap: break-word;
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 4px;
}

pre code {
  padding: 0;
  font-size: inherit;
  color: inherit; 
  white-space: pre-wrap;
  background-color: transparent;
  border: none;
  border-radius: 0;
}

code {
  padding: 2px 4px;
  color: inherit; 
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 4px;
  font-size: .9em;
}



blockquote,
.blockquote {
  padding: 10px 20px;
  margin: 0 0 20px;
  font-size: 1em;
  border-left: 5px solid #6c757d;
}

</style>
</head>

<body>
  <nav class="custom-navbar">
  <div class="container">
    
    <a href="/">Posts</a>
    
    <a href="/tags/">Tags</a>
    
    <a href="/about/">About</a>
    
    <a href="/index.xml">RSS</a>
    
  </div>
</nav>
  
  <div class="container">
    <article>
      <h1>[Python] Super easy molecular phylogenetic tree creation technique that I do not want to teach anyone</h1>
<p>
  <small class="text-secondary">
  
  
  Dec 1, 2019
  </small>
  

<small><code><a href="https://memotut.com/tags/python">Python</a></code></small>


<small><code><a href="https://memotut.com/tags/tree">tree</a></code></small>


<small><code><a href="https://memotut.com/tags/bioinformatics">bioinformatics</a></code></small>


<small><code><a href="https://memotut.com/tags/ete3">ete3</a></code></small>


<small><code><a href="https://memotut.com/tags/etetoolkit">etetoolkit</a></code></small>

</p>
<pre><code># Phylogenetic tree estimation
</code></pre>
<p>Introducing a super simple molecular phylogenetic tree creation method using <a href="http://etetoolkit.org/">ETE Toolkit</a>.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/57724/23c259d5-a789-c92a-9b25-52272b830db6.png" alt="NUP62.aa.fa.final_tree.png">
(Figure: From <a href="http://etetoolkit.org/documentation/ete-build/">http://etetoolkit.org/documentation/ete-build/</a>)</p>
<h2 id="tool-to-use">Tool to use</h2>
<ul>
<li>BLAST (sequence similarity search)</li>
<li>MAFFT (alignment)
<a href="https://mafft.cbrc.jp/alignment/server/">https://mafft.cbrc.jp/alignment/server/</a></li>
<li>trimAl (correction of alignment for phylogenetic tree construction), optional
<a href="http://trimal.cgenomics.org/">http://trimal.cgenomics.org/</a></li>
<li>RAxML (Physical tree calculation)
<a href="https://sco.h-its.org/exelixis/web/software/raxml/index.html">https://sco.h-its.org/exelixis/web/software/raxml/index.html</a></li>
<li><a href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a>or<a href="http://itol.embl.de/index.shtml">iTOL</a>(Drawingphylogenetictree)</li>
</ul>
<h2 id="rough-flow">Rough flow</h2>
<ol>
<li>BLAST sequence similarity search is performed on the 16S rRNA sequence (?) of the target species for which you want to see the phylogenetic relationship with related species against the sequence database (16S rRNA, RefSeq, nr, etc.), and the result is the target. Get the sequences of other species such as the genus to which the species belongs. *I am targeting bacteria 16S, but please use your own sequence.</li>
<li>Get the 16S rRNA sequence of the non-genus (to make it an outgroup of the phylogenetic tree).</li>
<li>Put all the sequences in ↑1.2 into a FASTA file and perform multiple alignment using MAFFT L-INS-i.</li>
<li>Adjust the alignment results using trimAl for phylogenetic tree construction (when calculating phylogenetic tree, remove unsuitable areas such as gaps). This step is optional and you do not have to do it.</li>
<li>Construct a maximum likelihood phylogenetic tree using RAxML. It also calculates the bootstrap value.</li>
<li>Read the constructed phylogenetic tree into FigTree and draw it. Or draw with iTOL (where the web base is better).</li>
<li>Fun (✿╹◡╹)v</li>
</ol>
<h2 id="keywords">Keywords</h2>
<p>Types of phylogenetic trees: unrooted tree, rooted tree,
Phylogenetic tree calculation method: NJ method, maximum likelihood method, Bayes method</p>
<h2 id="back-route">Back route</h2>
<p>A method to construct a phylogenetic tree in an instant using <a href="http://etetoolkit.org/">ETE Toolkit</a>.
It was so easy that I was really surprised. Actually, I wanted to use it as a paid information product, but to be honest, I was really worried. And I really thought about it seriously. I would like to make it public this time.</p>
<p>It is assumed that Python and Anaconda are included.
If it is not included/suspicious →
macOS: <a href="https://qiita.com/aical/items/2d066801a7464a676994">Note from installation of Homebrew to Python&rsquo;s Anaconda environment construction with pyenv</a>
Linux: <a href="https://qiita.com/aical/items/126128c3e8916ad1988f">pyenv in Linux environment Anaconda environment construction memo of Python</a></p>
<h3 id="environment">Environment</h3>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">$ pip install ete3 <span style="color:#75715e">#ETE toolkit installation</span>
$ conda install -c etetoolkit ete_toolchain <span style="color:#75715e"># install required tools</span>
</code></pre></div><h3 id="genealogical-tree-construction">Genealogical tree construction</h3>
<p>The grammar is</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">$ ete3 build -w workflow name -n input sequence file <span style="color:#f92672">(</span>before alignment<span style="color:#f92672">)</span> -o output destination directory name --clearall
</code></pre></div><p>Example:</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">$ ete3 build -w mafft_linsi-none-none-raxml_default -n input.fasta -o output_tree --clearall
</code></pre></div><p>Only this! This alone will do all the multiple alignment of sequences, trimming of many gap regions, and systematic estimation! !</p>
<p>The grammar of this workflow is <code>aligner-trimmer-tester-builder</code> separated by hyphens.
For example, if the workflow is
<code>mafft_linsi-none-none-raxml_default</code>
Then, the alignment is performed with mafft&rsquo;s L-INS-i algorithm, and the phylogenetic tree is constructed with RAxML.
If Paisen of the lab uses clustal for alignment! If you say <code>clustalo_default-none-none-raxml_default</code>, clustal omega (successor to clustal w) will be used for alignment.
If you want to include the process of trimming the gap area of the array using trimAl,
It should be <code>mafft_linsi-trimal01-none-raxml_default</code>.
If you want to include the bootstrap value when estimating the phylogenetic tree
Use <code>mafft_linsi-none-none-raxml_default_bootstrap</code> (it will take some time).</p>
<p>・Maybe it is safe to use <code>mafft_linsi-none-none-raxml_default_bootstrap</code> (sorry sorry)</p>
<p>Tools that can be used</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-bash" data-lang="bash">$ ete3 build apps
</code></pre></div><p>Will display it. <a href="http://etetoolkit.org/cookbook/ete_build_workflows.ipynb">Reference: Composing custom workflows</a></p>
<h3 id="result-file">Result file</h3>
<p>When you execute the phylogenetic tree construction command in the above example,
It creates various result files in <code>./output_tree/mafft_linsi-none-none-raxml_default</code>.
There are various files in it,
・Input.fasta.final_tree.png
→ Chara diagram that shows the phylogenetic tree and the schematic diagram of the aligned sequences together
・Input.fasta.final_tree.fa
→ aligned fasta file
・Input.fasta.final_tree.nw
→ Estimated phylogenetic tree file. If you load this into FigTree or iTOL, you can get a beautiful figure.</p>
<p><a href="http://etetoolkit.org/cookbook/">Reference: The ETE Cookbook</a></p>

    </article>
  </div>

  
  
  
  <script>
  window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
  ga('create', 'UA-169005401-1', 'auto');
  ga('send', 'pageview');
  </script>
  <script async src='https://www.google-analytics.com/analytics.js'></script>
  

  

  
<link rel="stylesheet" type="text/css" href="//cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.1.0/cookieconsent.min.css" />
<script src="//cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.1.0/cookieconsent.min.js"></script>
<script>
window.addEventListener("load", function(){
window.cookieconsent.initialise({
  "palette": {
    "popup": {
      "background": "#216942",
      "text": "#b2d192"
    },
    "button": {
      "background": "#afed71"
    }
  }
})});
</script>

</body>

</html>
