<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  
  <meta name="generator" content="Hugo 0.72.0" />

  
  <meta name="description" content="Some description">
  

  
  <link rel="apple-touch-icon" sizes="180x180" href="https://memotut.com/apple-touch-icon.png">

  
  <link rel="icon" type="image/png" sizes="32x32" href="https://memotut.com/favicon-32x32.png">

  
  <link rel="icon" type="image/png" sizes="16x16" href="https://memotut.com/favicon-16x16.png">

  
  <link rel="manifest" href="https://memotut.com/site.webmanifest">

  
  <link rel="mask-icon" href="https://memotut.com/safari-pinned-tab.svg" color="#5bbad5">

  <meta name="msapplication-TileColor" content="#da532c">

  <meta name="theme-color" content="#ffffff">

  
  <link rel="stylesheet" href="https://memotut.com/css/bootstrap.min.css" />

  
  <title>[Python] ChIP-seq analysis starting from scratch | Memo Tut</title>
  

  <style>
body {
  min-width: 300px;
}

.custom-navbar {
  margin-bottom: 1em;
  height: 60px;
}

.custom-navbar a {
  display: inline-block; 
  padding: 18px 0;
  margin-right: 1em; 
  font-weight: bold; 
}

.custom-navbar a:hover,
.custom-navbar a:focus {
  text-decoration: none; 
}

@media print {
  .custom-navbar {
    display: none;
  }
}

article {
  padding-bottom: 1em;
}

img {
  max-width: 100%;
}


body {
  background-color: #fff;
}



body {
  color: #212529;
}



a {
  color: #007bff;
}



a:hover,
a:focus {
  color: #0056b3;
}



.custom-navbar {
  background-color: #212529;
}



.custom-navbar a {
  color: rgba(255, 255, 255, 0.75);
}



.custom-navbar a:hover,
.custom-navbar a:focus {
  color: rgba(255, 255, 255, 1);
}



.container {
  max-width: 800px;
}



pre {
  display: block;
  padding: 9.5px;
  word-break: break-all;
  word-wrap: break-word;
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 4px;
}

pre code {
  padding: 0;
  font-size: inherit;
  color: inherit; 
  white-space: pre-wrap;
  background-color: transparent;
  border: none;
  border-radius: 0;
}

code {
  padding: 2px 4px;
  color: inherit; 
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 4px;
  font-size: .9em;
}



blockquote,
.blockquote {
  padding: 10px 20px;
  margin: 0 0 20px;
  font-size: 1em;
  border-left: 5px solid #6c757d;
}

</style>
</head>

<body>
  <nav class="custom-navbar">
  <div class="container">
    
    <a href="/">Posts</a>
    
    <a href="/tags/">Tags</a>
    
    <a href="/about/">About</a>
    
    <a href="/index.xml">RSS</a>
    
  </div>
</nav>
  
  <div class="container">
    <article>
      <h1>[Python] ChIP-seq analysis starting from scratch</h1>
<p>
  <small class="text-secondary">
  
  
  Dec 25, 2019
  </small>
  

<small><code><a href="https://memotut.com/tags/python">Python</a></code></small>


<small><code><a href="https://memotut.com/tags/bioinformatics"> bioinformatics</a></code></small>


<small><code><a href="https://memotut.com/tags/chip-seq"> ChIP-seq</a></code></small>


<small><code><a href="https://memotut.com/tags/homer"> Homer</a></code></small>

</p>
<pre><code>#Introduction
</code></pre>
<p>ChIP-seq (chromatin immunoprecipitation followed by sequencing) is a method to comprehensively measure at which position on the genome the binding and histone modification of a specific transcription factor and how often it occurs. You will be able to start from the environment construction, analyze using the data of the paper, and finally see the results in the genome browser and make peak calls ~. Let&rsquo;s start with the environment construction immediately!</p>
<h1 id="environment">Environment</h1>
<h2 id="install-miniconda3">Install miniconda3</h2>
<p>First, install the package manager, miniconda, which makes it easy to install and manage tools when building your environment.</p>
<p>For Linux</p>
<pre><code>wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh

bash Miniconda3-latest-Linux-x86_64.sh
</code></pre><p>For Mac</p>
<pre><code>wget https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh

bash Miniconda3-latest-MacOSX-x86_64.sh
</code></pre><p>When you have done this, press ENTER when prompted and type <code>yes</code>. You can type <code>yes</code> in any question. Close the terminal once you are done. Start it and add a channel to miniconda next time, so please execute the following._ Please be sure to execute in this order __.</p>
<pre><code>conda config --add channels defaults
conda config --add channels conda-forge
conda config --add channels bioconda
</code></pre><p>If you are told there is no conda here, please refer to help 1 below.</p>
<h2 id="install-required-tools">Install required tools</h2>
<p>Now, we will install the necessary tools using miniconda. This time, install the following tools.</p>
<ul>
<li><a href="https://github.com/ncbi/sra-tools">SRA Toolkit</a> Tools used to download sequence data from SRA</li>
<li><a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a> Adapter trimming tool</li>
<li><a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">fastQC</a> Tool for QC of fastq files</li>
<li><a href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">Bowtie2</a> Mapping tool</li>
<li><a href="http://broadinstitute.github.io/picard/">Picard</a> PCR duplicate removal tool</li>
<li><a href="http://samtools.sourceforge.net/">samtools</a> Tools used in various places</li>
<li><a href="https://deeptools.readthedocs.io/en/develop/">deepTools</a> Tools used for data correction</li>
<li><a href="http://homer.ucsd.edu/homer/ngs/">HOMER</a> Various tools</li>
</ul>
<p>Install using miniconda&rsquo;s <code>conda</code> command as follows.</p>
<pre><code>conda install sra-tools
conda install trimmomatic
conda install fastqc
conda install bowtie2
conda install picard
conda install samtools
conda install deeptools
conda install homer
</code></pre><p>Answer <code>y</code> for all <code>Proceed ([y]/n)?</code>.
The environment is now ready! I will touch the data from now on, but in order to avoid confusion in this article ** I wrote assuming that all commands are done in the same directory **. Let&rsquo;s get the data ~.</p>
<p>#Download sequence data</p>
<h2 id="check-if-there-is-data">Check if there is data</h2>
<p>Download the sequence data from <a href="https://www.ncbi.nlm.nih.gov/sra">SRA(Sequence Read Archive)</a>.Thistime,<a href="https://ChIP-seqdataofMed1ofmouseEScellsfrom%5Bwww.nature.com/articles/nature09380">KageyMH<em>etal.,</em></a>(GSM560348)and[Creyghton<em>etal.,</em>](<a href="https://www.pnas.org/content/107(50/21931.long)(GSM594579)alsousesChIP-seqdataofH3K27AcofmouseEScells.Inputsequencedata(GSM560357)isalsoobtainedfromthesamepaperasMed1.ThenumberinsideistheGEOAccessionnumberofeachdata.ItisthenumberrequiredtopagethesedatabyGeneExpressionOmnibus(GEO)">https://www.pnas.org/content/107(50/21931.long)(GSM594579)alsousesChIP-seqdataofH3K27AcofmouseEScells.Inputsequencedata(GSM560357)isalsoobtainedfromthesamepaperasMed1.ThenumberinsideistheGEOAccessionnumberofeachdata.ItisthenumberrequiredtopagethesedatabyGeneExpressionOmnibus(GEO)</a> of NCBI.</p>
<p>Now, before downloading the data, let&rsquo;s first confirm the existence of these data on a browser. The data of Med1 is explained as an example. First, <a href="https://www.ncbiOpen.nlm.nih.gov/geo/">GEO site</a> in your browser. Type <code>GSM560348</code> in the search window circled in red in the image below.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/0897f373-0e7c-5c96-75b5-79d412396413.png" alt="GEO1.PNG"></p>
<p>Please read carefully as various information about this data is written like the image below.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/40bf9c1e-1872-a869-417a-446db971722c.png" alt="GEO2.PNG"></p>
<p>Then, click the number on the right of the word SRA at the bottom of this page.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/427e31df-8e2a-c437-082f-c2962d04e763.png" alt="GEO3.PNG"></p>
<p>Then you will go to the following page.The number enclosed in red in the image is the SRR number, which is necessary for downloading with the access number of this data.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/ba16dca8-8471-e787-3de0-2445e219e78b.png" alt="GEO4.PNG"></p>
<p>These numbers are written somewhere in the paper so that you can see if there is data.</p>
<h2 id="download-data-from-sra">Download data from SRA</h2>
<p>Use the sratoolkit command <code>fastq-dump</code>. It is very easy to use and if the data is single ended</p>
<pre><code>fastq-dump SRR number of data to download
</code></pre><p>For paired ends,</p>
<pre><code>fastq-dump SRR number of the data you want to download --split-files
</code></pre><p>This should download the fastq file, which is the raw data output by the sequencer, to the directory where you ran this command.</p>
<p>Then I will download this data.</p>
<pre><code>fastq-dump SRR058988 #Med1
fastq-dump SRR066767 #H3K27Ac
fastq-dump SRR058997 # input
</code></pre><p>You can also write it down as follows and download it at once.</p>
<pre><code>fastq-dump SRR058988 SRR066767 SRR058997
</code></pre><p>This process downloads the .sra file, which is a compressed file format specific to SRA, and converts it to a fastq file. It will take time, so please wait patiently. This time I will use the bioinformatics of How about general commentary? ?</p>
<p><a href="https://laborify.net/2019/11/30/michida-bioinformatics/">https://laborify.net/2019/11/30/michida-bioinformatics/</a></p>
<p>After downloading, the file name is attached with the SRR number, so rename it and make it easier to understand.</p>
<pre><code>mv SRR058988.fastq Med1.fastq
mv SRR066767.fastq H3K27Ac.fastq
mv SRR058997.fastq Input.fastq
</code></pre><h1 id="adapter-trimming">Adapter trimming</h1>
<p>Now, we will use Trimmomatic to clean the sequence results.</p>
<pre><code>trimmomatic SE -threads 4 -phred33 Med1.fastq Med1_trimmed.fastq ILLUMINACLIP:$HOME/miniconda3/share/trimmomatic/adapters/TruSeq3-SE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN: 36
</code></pre><p>It&rsquo;s long, but it doesn&rsquo;t include line breaks to prevent accidents due to line breaks. Immediately after <code>trimmomatic</code>, specify whether the data to be analyzed is single-ended (SE) or paired-ended (PE). With the next <code>-threads</code> Specify the number of threads to use. Do not enter <code>-phred33</code>. Be sure to enter it. After that, enter the name of the file to be trimmed and the name of the file after trimming.What is written after ILLUMINACLIP is the location of the adapter array information, and it should be in such a place under the miniconda3 directory. Rewrite according to the location of your miniconda3. (Do not do anything special when installing) It should be in your home directory like this.) Also, 2:30:10 represents the number of allowable mismatches, palindrome clip threshold, and simple clip threshold respectively. Basically, I think that it is okay not to mess with here. Also, <code>LEADING:3</code> and <code>TRAILING:3</code> mean to delete the bases with quality score less than 3 from the beginning and end of the read respectively. <code>SLIDINGWINDOW:4:15</code> means every 4 bp. It means to look at the average quality score and remove the part where it is less than 15. And the last <code>MINLEN:36</code> means to exclude those with read length less than 36 from the analysis. These settings are from &ldquo;Trimmomatic page&rdquo; (<a href="http://www.usadellab.org/cms/?page=trimmomatic">http://www.usadellab.org/cms/?page=trimmomatic</a>) &ldquo;Quick start&rdquo;.
When done, a file named Med1_trimmed.fastq will be generated. Please execute it with the same options for the remaining two data.</p>
<p>#QC of fastq file after trimming</p>
<p>Quality control of fastq files after trimming is done using fastQC.</p>
<pre><code>fastqc --threads 4 --nogroup -o .Med1_trimmed.fastq
</code></pre><p>Write the number of threads with <code>--threads</code> immediately after <code>fastqc</code>. If you write the next <code>--nogroup</code>, the lead at the 3&rsquo;end is also analyzed. The result is output to <code>-o</code>. Finally, write the name of the file to be QC&rsquo;d.</p>
<p>A file called <code>Med1_trimmed_fastqc.html</code> will be created in the directory specified in <code>-o</code>, so let&rsquo;s open it with a browser.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/c354cfad-2864-b917-02fa-7ce8d6c7bdef.png" alt="fastqc.PNG"></p>
<p>The quality is OK if the summary on the left is almost green check. This data is too beautiful &hellip; Each index is explained in <a href="https://bi.biopapyrus.jp/rnaseq/qc/fastqc.html">https://bi.biopapyrus.jp/rnaseq/qc/fastqc.html</a> Thank you for your continued support of bioinformatics, this site.</p>
<p>Please do the same for the remaining two data.</p>
<p>#Mapping</p>
<p>We will finally use Bowtie 2 to map! And before that, we have to build the index of the genome. In other words, we need to prepare the reference genome required for mapping.</p>
<p>Download the whole array of mm10 from <a href="https://hgdownload.soe.ucsc.edu/downloads.html">UCSC page</a> with <code>wget</code> command. Create a folder called <code>ref_genome</code> and drop it there. The</p>
<pre><code>mkdir ref_genome
cd ref_genome
wget http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/chromFa.tar.gz
tar -xzf chromFa.tar.gz
</code></pre><p>In addition, we will discard random and unknown sequences and mitochondrial genome (chrM.fa) this time.</p>
<pre><code>rm *random.fa
rm chrUn*
rm chrM.fa
</code></pre><p>Make the remaining files into one file called <code>mm10.fa</code> using <code>cat</code>.</p>
<pre><code>cat *.fa &gt;mm10.fa
</code></pre><p>We will create a directory mm10_index in the same hierarchy as <code>ref_genome</code> and save the index there.</p>
<pre><code>cd .. #if you are in ref_genome now
mkdir mm10_index
bowtie2-build -f ./ref_genome/mm10.fa ./mm10_index/mm10_index --threads 4
</code></pre><p><code>bowtie2-build</code> is a command to make Bowtie2 index. Write the path of the location of the genome sequence in the <code>-f</code> option and then the path to the index. Specify the number of threads with <code>--threads</code>. You can. In this case, six files <code>mm10_index*.bt2</code> should be created in the directory <code>mm10_index</code> created earlier. You only need to do this once.</p>
<p>It takes a lot of time. Let&rsquo;s use this time to study the R language that is often used for statistical analysis in bioinformatics. (Thank you!)</p>
<p><a href="https://qiita.com/roadricefield/items/001c882f84dd093f4407">https://qiita.com/roadricefield/items/001c882f84dd093f4407</a></p>
<p>I won&rsquo;t use R in this article, but&hellip;</p>
<p>*It takes a really long time. It&rsquo;s a good level to hit the command just before returning home and continue when tomorrow comes.</p>
<p>&hellip;&hellip;&hellip;&hellip;&hellip;&hellip;&hellip;&hellip;.</p>
<p>Good morning everybody! I forgot to specify <code>--threads</code> so it took 7 hours! Now let&rsquo;s map.</p>
<pre><code>bowtie2 -p 4 -x ./mm10_index/mm10_index -U Med1_trimmed.fastq -S Med1.sam
</code></pre><p><code>-p</code> is the number of threads, <code>-x</code> is the index, <code>-U</code> is the fastq file to be mapped, and <code>-S</code> is the output file name. It is output as a sam file once. By the way, mapping takes a long time. It&rsquo;s about 30 minutes.</p>
<p>When you&rsquo;re done, use samtools to convert the sam file to a bam file.</p>
<pre><code>samtools view -b -o Med1.bam Med1.sam
</code></pre><p>Mapping is now complete! Please do the same for the remaining two data.</p>
<p>#PCR duplicate removal (optional)</p>
<p>Finally, use Picard to remove the PCR duplicate. This is not always necessary.</p>
<pre><code>samtools sort Med1.bam &gt;Med1_sorted.bam #bic files need to be sorted in order to use Picard.

picard MarkDuplicates I=Med1_sorted.bam O=Med1_rm_dups.bam M=Med1_report.txt REMOVE_DUPLICATES=true
</code></pre><p>Write the name of the bam file where you want to remove PCR duplicate in <code>I</code>, the name of the file after removing duplicate in <code>O</code>, and a report that is a summary of the calculation results, so write that name in <code>M</code>. The</p>
<p>Do the same for the other two data.</p>
<p>Since the directory I&rsquo;m working on is almost full, I will organize the data here.The data of Med1 ChIP-seq is in the directory <code>Med1_data</code> and the data of H3K27Ac ChIP-seq is in the directory <code>H3K27Ac_data</code>. Let&rsquo;s move the data to a directory called <code>Input_data</code>.</p>
<pre><code>mkdir Med1_data
mv Med1* Med1_data

mkdir H3K27Ac_data
mv H3K27Ac* H3K27Ac_data

mkdir Input_data
mv Input* Input_data
</code></pre><h1 id="observe-results-in-genome-browser">Observe results in Genome Browser</h1>
<h2 id="create-a-bigwig-file">Create a bigWig file</h2>
<p>First of all, let&rsquo;s convert the bam file into a format called bigWig that is easy to see in the genome browser.To do this, use the deepTools <code>bamCoverage</code>.In the bigWig format, the correction value of the number of reads in the determined bin width is calculated for the whole genome. In other words, the signal intensity of ChIP-seq at each bin on the genome is calculated. In order to make this, it is necessary to first create a bam index file bam.bai file, so use samtools Let&rsquo;s make it.</p>
<pre><code>samtools index Med1_data/Med1_rm_dups.bam
</code></pre><p>Doing this will create an index file called <code>Med1_rm_dups.bam.bai</code>.</p>
<p>Now run <code>bamCoverage</code>. Be sure to put the bam file and its bam.bai file in the same directory and run it.</p>
<pre><code>bamCoverage -b Med1_data/Med1_rm_dups.bam -p 4 --normalizeUsing RPGC --effectiveGenomeSize 2652783500 --binSize 1 -o Med1_data/Med1.bigwig
</code></pre><p>Write the name of the bam file to be converted into a bigWig file in <code>-b</code>. <code>-p</code> is the number of threads. In <code>--normalizeUsing</code>, select the type of correction value to be calculated in each bin. <code>RPKM</code>, <code> You can select CPM</code>, <code>BPM</code>, <code>RPGC</code>, or <code>None</code>. If you select <code>None</code>, the number of reads contained in the bin will be the value in that bin. For <code>--effectiveGenomeSize</code> Enter the length (bp) of the part that can be mapped.If it is mm10 (also known as GRCm38), it is <code>2652783500</code>. (Reference <a href="https://deeptools.readthedocs.io/en/latest/content/feature/">https://deeptools.readthedocs.io/en/latest/content/feature/</a> effectiveGenomeSize.html) In <code>--binSize</code>, enter the bin length (bp) to be used in the calculation.In <code>-o</code>, write the name of the output file.</p>
<p>The calculation takes time, so let&rsquo;s install Genome Browser in the meantime.</p>
<h2 id="install-igv-integrative-genomics-viewer">Install IGV (Integrative Genomics Viewer)</h2>
<p>Genome Browser is a tool to visualize the sequence result. You often see the visualization of 〇〇-seq signal at a specific position in the genome, right? That&rsquo;s it. Let&rsquo;s install it now<img src="https://software.broadinstitute.org/software/igv/download" alt="IGV download page"> Please download your OS installer. IGV is GUI (Graphic comes out and operates with mouse and keyboard. If you are a Windows user using WSL, please select the Windows version here. Start the downloaded installer and follow the instructions to install it. Then, the following IGV shortcut will be created on your desktop. The</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/bd3c8a57-c96e-f1c9-255e-165de62c9d4c.png" alt="IGV1.PNG"></p>
<p>Double-click this to start it (it takes about 30 seconds to start). After starting, the following window will appear.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/558efcf8-cfc4-96ac-c6b7-67b7e482b544.png" alt="IGV2.PNG"></p>
<p>Now, hg19 is loaded, so let&rsquo;s download and load mm10. Click the down arrow in the above screen in red, and &ldquo;More&hellip;&rdquo; will appear. Click it. Then click &ldquo;Mouse mm10&rdquo; and check &ldquo;Download Sequence&rdquo; at the bottom left and click &ldquo;OK&rdquo;. Now mm10 download will start.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/138e1746-c4a6-9f2f-abbd-90cf22d53fd0.png" alt="IGV3.PNG"></p>
<p>It&rsquo;s about time the <code>bamCoverage</code> is over&hellip;?</p>
<p>When you are done, drag and drop <code>Med1.bigwig</code> onto the IGV window.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/7e3de502-d572-1050-b56e-af8a26a5d76d.png" alt="IGV4.PNG"></p>
<p>Is the profile of Med1 ChIP-seq displayed as in the image above? In this state, we are looking at all the chromosomes in a bird&rsquo;s-eye view, and since we do not know the details, we input various gene names in the red search window. And jump to that gene body location. Below is just one example.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/e18a9200-49d3-05cf-c7a5-65f31c0232e0.png" alt="IGV_Klf4.PNG"></p>
<p><em>Klf4</em> is a gene that is important for maintaining stemness, but it is clear that Med1 that is a protein that activates transcription is bound in mouse ES cells. The track is being pulled up. If the vertical scale of the display is incorrect, right-click on the place where the name &ldquo;Med1.bigwig&rdquo; is written and click &ldquo;Autoscale&rdquo;.</p>
<p>Create bigwig for the remaining two data and check it with IGV.</p>
<p>#Peak call</p>
<p>Then we will make a peak call that detects the peak of the signal based on statistical criteria. This time we will use <code>findPeaks</code> of HOMER. Other than this, we also use <a href="https://github.com/taoliu/MACS">MACS2</a>. If you are interested, please compare the results.</p>
<p>Now, before doing a <code>findPeaks</code>, you need to convert the bam file into a TagDirectory that HOMER can handle, using HOMER&rsquo;s <code>makeTagDirectory</code>.</p>
<pre><code>makeTagDirectory Med1_data/Med1_TagDir -single Med1_data/Med1_rm_dups.bam
</code></pre><p>Write the name of the TagDirectory to be created immediately after <code>makeTagDirectory</code>, then list the options and write the name of the bam file at the end. For this option, specify only the <code>-single</code> option that cleans the inside of TagDirectory. See <a href="http://homer.ucsd.edu/homer/ngs/tagDir.html">http://homer.ucsd.edu/homer/ngs/tagDir.html</a> for options to. Create TagDirectory for the remaining two data in the same way.</p>
<p>Now run <code>findPeaks</code>.</p>
<pre><code>findPeaks Med1_data/Med1_TagDir -style factor -o auto -i Input_data/Input_TagDir
</code></pre><p>Immediately after <code>findPeaks</code>, write the name of the TagDirectory that makes the peak call.For the <code>-style</code> option, consider Med1 as a transcription factor and enter <code>factor</code>. <code>-o</code> option Write the location to write the result to, but if you set it as <code>auto</code>, it will be saved in the TagDirectory that makes the peak call. Also, in the <code>-i</code> option, write the TagDirectory name of the input data. Input data When there is no input, if you do not enter the <code>-i</code> option itself, the calculation will be performed without input.</p>
<p>Next, perform a peak call of H3K27Ac ChIP-seq data.</p>
<pre><code>findPeaks H3K27Ac_data/H3K27Ac_TagDir -style histone -o auto -i Input_data/Input_TagDir
</code></pre><p>Since H3K27Ac is histone modified data, set <code>-style</code> option to <code>histone</code>.</p>
<p>Let&rsquo;s take a look at the peak call result file. I think there is a file <code>peaks.txt</code> in the TagDirectory (H3K27Ac has the name <code>regions.txt</code>). Since it is text, it is easy to see when opened in Excel.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/044d2933-eaa7-8e5b-2e72-e8fcf86d9cbe.png" alt="findPeaks.PNG"></p>
<p>After various information is written in the upper part, the information of the peak like the lower half of the image is written.If you look at the columns of <code>chr</code>, <code>start</code>, and <code>end</code>, you can see the genome of each peak. You can see the position above.The <code>Normalized Tag Count</code> is the signal intensity of each peak.Divided by 10 is rpm (Reads per Million).For more details about the function of <code>findPeaks</code>, see http://homer. See ucsd.edu/homer/ngs/peaks.html.</p>
<p>Next, let&rsquo;s take a look at this result in IGV together with the bigwig file.The bed file is a good way to make IGV read the peak information easily.The bed file is very simple and simple. It is a text file like the image below that records the position information of the peak consisting of three lines of <code>chr</code>, <code>start</code>, and <code>end</code> from the left side. (There may be cases.)</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/cee9e82b-8125-08e7-cb16-cf749c502395.png" alt="bed.PNG"></p>
<p>You can create it from the <code>peaks.txt</code> output of <code>findPeaks</code> with the following command.</p>
<pre><code>sed'/^#/d' Med1_data/Med1_TagDir/peaks.txt | awk -v'OFS=\t''{print $2, $3, $4}' &gt;Med1_data/Med1_TagDir/peaks.bed
</code></pre><p>Try dragging and dropping the <code>peaks.bed</code> you just made to the IGV.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/6e48bd58-d569-092e-73a0-5399eec51cce.png" alt="peak.PNG"></p>
<p>You should see the peak called position like this. Easy to understand&hellip;</p>
<p>#Binding motif analysis</p>
<p>Transcription factors bind to a certain sequence (binding motif sequence), so when a peak region is given, by examining what sequences are enriched in them, what transcription factors bind to them For example, by performing binding motif analysis in the ChIP-seq peak region of Med1, it is possible to predict what kind of transcription factor binds to the place where Med1 binds. Let&rsquo;s do this using HOMER&rsquo;s <code>findMotifsGenome.pl</code>.</p>
<p>This also requires preliminary preparation and it is necessary to install the genome in HOMER.</p>
<pre><code>perl $HOME/miniconda3/share/homer-4.10-0/configureHomer.pl -install mm10
</code></pre><p>Install mm10 with <code>configureHomer.pl</code> in the above location under the miniconda3 directory. The above is when miniconda3 is in the home directory. Also, for homer-4.10-0, you can use <code>conda install Please note that it may change depending on when HOMER is installed with homer</code> and the environment. (Because the HOMER version may be different.) WSL people who did not work here help below. See 2.</p>
<p>Now run <code>findMotifsGenome.pl</code>.</p>
<pre><code>findMotifsGenome.pl Med1_data/Med1_TagDir/peaks.txt mm10 Med1_data/Med1_motif -size given -p 4
```Immediately after `findMotifsGenome.pl`, write the result file of `findPeaks`. (If you want to use the results of other peak callers, please refer to help 3 below.) Then write the version of genome. In addition, write the name of the directory to save the results.The calculation will be performed for the area centered on the center of each peak of integer length input to the `-size` option. And `-size 100`, the calculation is performed within +/- 50 bp from the center of each peak. This time, the calculation is performed within the actual range of each peak area passed to the program as `-size given`. Enter the number of threads to use for `-p`.

When the calculation is over, a directory called `Med1_motif` will be created in `Med1_data`, so let's take a look inside. The noteworthy is `knownResults.html`. Open this in a browser. If you get an error, please refer to Help 2 below.

![Motif.PNG](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/4db9bd02-a008-3315-f59e-4f09774edb1f.png)
 

In the case of this calculation, the p-value indicates which motif sequence on HOMER database was enriched in the input peak region for the automatically prepared random region. They are displayed in ascending order. Since the calculation is performed on a random region, the result changes slightly every time.The results show that the peak region of Med1 is used to maintain the stemness of KLF, OCT, SOX, etc. Many important transcription factor motifs are present, suggesting that Med1 and these transcription factors co-localize.

Apart from this, in `homerResults.html`, there are many sequences that exist in the input peak region and are similar to the motif sequence on the HOMER database. Basically, you can check `knownResults.html`. It would be nice.

For more information about `findMotifsGenome.pl`, see http://homer.ucsd.edu/homer/ngs/peakMotifs.html.

#Check the overlap between peak areas

Finally, I will introduce a method to investigate the overlap between peaks. To do this, use HOMER's `mergePeaks`. This time, we will investigate the overlap between the peak areas of Med1 ChIP-seq and H3K27Ac ChIP-seq.

</code></pre><p>mergePeaks -d given Med1_data/Med1_TagDir/peaks.txt H3K27Ac_data/H3K27Ac_TagDir/regions.txt -prefix mergePeaks -venn venn.txt</p>
<pre><code>
When `-d` is set to `given`, the overlap of the input peak areas is calculated as it is. The peak areas to be input are written and arranged. There may be three or more. Preserve as `-prefix XXX` And the file name starts with `XXX`, the combined area of overlapping peaks and the area that exists only in one peak area are output separately. If you set it as `-venn YYY.txt`. Creates a table that summarizes the number of overlapping peak areas, `YYY.txt`, and the number of peak areas in only one of them in order to draw a Venn diagram. For details on options, etc. http://homer.ucsd See .edu/homer/ngs/mergePeaks.html.


When this command is executed, only merged_e3_27_H3K27Ac_data_H3K27Ac_data_H3K27Ac_data. , A region that exists only in the peak of Med1 ChIP-seq, a region in which overlapping of the peak regions of Med1 ChIP-seq and H3K27Ac ChIP-seq are combined, it is a table for drawing a Venn diagram.

Let's draw a Venn diagram with matplotlib in Python. Since we use a package called matplotlib_venn,

</code></pre><p>conda install matplotlib-venn</p>
<pre><code>
Now you can do anything, open the editor and write the code below.

```python
from matplotlib import pyplot as plt
from matplotlib_venn import venn2

Open #venn.txt and find the number of peaks only in Med1 ChIP-seq,
#H3K27Ac Check the number of peaks existing only in ChIP-seq and the number of overlapping peaks.

venn2(subsets=(770, 25254, 2738), set_labels = (&quot;Med1&quot;, &quot;H3K27Ac&quot;))
#subsets=(Med1 only, H3K27Ac only, overlapping peaks)

plt.savefig(&quot;./venn.png&quot;)
</code></pre><p>After writing it, save it with a name such as <code>venn_plot.py</code> and execute the following command in the saved location.</p>
<pre><code>python venn_plot.py
</code></pre><p>Then a file called <code>venn.png</code> will be created in that directory, so open it.</p>
<p><img src="https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/477619/f7617322-7215-9181-097d-feb213a0dd12.png" alt="venn.png"></p>
<p>About 80% of the peak of ChIP-seq of Med1 which is a protein that activates transcription overlaps the peak of ChIP-seq of H3K27Ac which is also a transcriptional activity marker. However, the number of peaks of H3K27Ac ChIP-seq is large. By the way, please compare the two data with IGV.</p>
<h1 id="at-the-end">At the end</h1>
<p>Thank you for reading so far. The analysis introduced after peak call is only a part of ChIP-seq analysis. Everyone is already able to perform mapping and peak call Use HOMER, R, Python, etc. to perform an analysis tailored to your purpose. Information about bioinformatics, including this article, is now available on the Internet. It is also possible to analyze ATAC-seq (Assay for Transposase-Accessible Chromatin Sequencing) comprehensively. We hope this article will be useful for your research. If you have any questions, please do as much as possible. I&rsquo;m sorry, so I&rsquo;m waiting for comments!</p>
<h1 id="help">help</h1>
<h2 id="1-the-conda-command-does-not-work">1. The <code>conda</code> command does not work!</h2>
<p>Most likely it&rsquo;s because the path isn&rsquo;t there after you installed miniconda3. Please do the following:</p>
<pre><code>cd # move to home directory

vim .bash_profile # write path .bash_profile open in vim
</code></pre><p>When it opens, press the esc key first, then the i key. Now you can edit in Insert mode. Make sure you type the following correctly.</p>
<pre><code>PATH=$PATH:~/miniconda3/bin
</code></pre><p>Write the path to <code>miniconda3/bin</code> after <code>PATH=$PATH:</code>. The above is if miniconda is in your home directory. If you make sure that you made no mistakes, press esc again. Type wq` and press enter.</p>
<p>Then restart the terminal</p>
<pre><code>source .bash_profile
</code></pre><p>Now, <code>conda</code> should work through the path.</p>
<h2 id="2-configurehomerpl--install-mm10-does-not-work">2. <code>configureHomer.pl -install mm10</code> does not work!</h2>
<p>Perhaps only WSL people can get this error because it doesn&rsquo;t have the necessary commands installed to run <code>configureHomer.pl</code>. Do all of the following:</p>
<pre><code>which gcc
which g++
which make
which perl
which zip
which gzip
which wget
</code></pre><p>this house</p>
<pre><code>/usr/bin/make
</code></pre><p>If the path to the command was not displayed, such as</p>
<pre><code>When sudo apt install zip #zip is not included
</code></pre><p>Please install again.</p>
<pre><code>perl $HOME/miniconda3/share/homer-4.10-0/configureHomer.pl -install mm10
</code></pre><p>Try running.</p>
<h2 id="3-using-bed-files-created-by-programs-other-than-homer-with-homer">3. Using bed files created by programs other than HOMER with HOMER</h2>
<p>When using a bed file created by a program other than HOMER with HOMER, it is safe to convert it into a HOMER bed file beforehand. To do this, use <code>bed2pos.pl</code> of HOMER.</p>
<pre><code>bed2pos.pl (bed file you want to convert) &gt;Converted_file.hb
</code></pre><p>HOMER bed files have the extension &ldquo;hb&rdquo;.</p>
<h1 id="references">References</h1>
<ul>
<li>
<p><a href="https://www.nature.com/articles/nature09380">Kagey MH <em>et al.,</em> Mediator and cohesin connect gene expression and chromatin architecture. Nature 467, 430–435 (2010).</a></p>
</li>
<li>
<p><a href="https://www.pnas.org/content/107/50/21931.long">Creyghton <em>et al.,</em> Histone H3K27ac separates active from poised enhancers and predicts developmental state.PNAS 107, 21931–21936 (2010).</a></p>
</li>
<li>
<p><a href="http://rnakato.hatenablog.jp/entry/2017/07/06/110926-">http://rnakato.hatenablog.jp/entry/2017/07/06/110926-</a> <a href="https://bi.biopapyrus.jp/rnaseq/qc/trimmomatic.html">https://bi.biopapyrus.jp/rnaseq/qc/trimmomatic.html</a></p>
</li>
<li>
<p><a href="http://www.usadellab.org/cms/?page=trimmomatic">http://www.usadellab.org/cms/?page=trimmomatic</a></p>
</li>
<li>
<p><a href="https://bi.biopapyrus.jp/rnaseq/qc/fastqc.html">https://bi.biopapyrus.jp/rnaseq/qc/fastqc.html</a></p>
</li>
<li>
<p><a href="https://bi.biopapyrus.jp/rnaseq/mapping/bowtie2/">https://bi.biopapyrus.jp/rnaseq/mapping/bowtie2/</a></p>
</li>
</ul>

    </article>
  </div>

  
  
  
  <script>
  window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
  ga('create', 'UA-169005401-1', 'auto');
  ga('send', 'pageview');
  </script>
  <script async src='https://www.google-analytics.com/analytics.js'></script>
  

  

  
<link rel="stylesheet" type="text/css" href="//cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.1.0/cookieconsent.min.css" />
<script src="//cdnjs.cloudflare.com/ajax/libs/cookieconsent2/3.1.0/cookieconsent.min.js"></script>
<script>
window.addEventListener("load", function(){
window.cookieconsent.initialise({
  "palette": {
    "popup": {
      "background": "#216942",
      "text": "#b2d192"
    },
    "button": {
      "background": "#afed71"
    }
  }
})});
</script>

</body>

</html>
